compiler/src/iree/compiler/Preprocessing/Common/ConvertConvFilterToChannelsLast.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Preprocessing/Common/Passes.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/LogicalResult.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-preprocessing-convert-conv-filter-to-channels-last"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler::Preprocessing {
+
+#define GEN_PASS_DEF_CONVERTCONVFILTERTOCHANNELSLASTPASS
+#include "iree/compiler/Preprocessing/Common/Passes.h.inc"
+
+// Utility function to swap the last two dimensions.
+static AffineMap constructFilterMap(AffineMap map, ArrayRef<int64_t> perm) {
+  unsigned numDims = map.getNumDims();
+  ArrayRef<AffineExpr> mapResults = map.getResults();
+  SmallVector<AffineExpr, 4> exprs;
+  for (int i = 0; i < perm.size(); ++i) {
+    exprs.push_back(mapResults[perm[i]]);
+  }
+  return AffineMap::get(numDims, map.getNumSymbols(), exprs, map.getContext());
+}
+
+// Utility function to create a transpose operation.
+static Value createTransposeOp(PatternRewriter &rewriter, Location loc,
+                               Value tensor, ArrayRef<int64_t> perm) {
+  SmallVector<OpFoldResult, 4> dimSizes =
+      tensor::getMixedSizes(rewriter, loc, tensor);
+  // Dim index of H, W, F, C
+  SmallVector<OpFoldResult, 4> transposeResultDims;
+  for (int i = 0; i < perm.size(); ++i) {
+    transposeResultDims.push_back(dimSizes[perm[i]]);
+  }
```

Comment:
Few nits:

- We don't need this comment. In this method, there is no convolution context.
- Instead of having the loop, you can reuse [the applyPermutationToVector method in upstream](https://github.com/llvm/llvm-project/blob/0a9c08c59ba61e727e9dee6d71883d9106963442/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L220-L229)

Reply:
Nice suggestion! How did you know such useful utility functions in `IndexingUtils.h`?